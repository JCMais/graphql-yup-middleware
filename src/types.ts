import * as Yup from 'yup';
import { GraphQLResolveInfo } from 'graphql';

export type YupMiddlewareErrorContext<TContext = any, TArgs = any> = {
  root: any;
  args: TArgs;
  context: TContext;
  info: GraphQLResolveInfo;
};

export type YupMiddlewareOptions = {
  errorPayloadBuilder?: (
    error: Yup.ValidationError,
    errorContext: YupMiddlewareErrorContext,
  ) => any;
  shouldTransformArgs?: boolean;
  yupOptions?: Parameters<Yup.BaseSchema['validate']>[1];
};

export type YupMiddlewareFieldValidationError = {
  field: string;
  errors: string[];
};

export type YupMiddlewareDefaultError = {
  message: string;
  details: YupMiddlewareFieldValidationError[];
};

export interface GraphQLExtensionsYupMiddleware<
  _TSource,
  _TContext,
  _TArgs extends Record<string, any> = Record<string, any>
> {
  validationOptions?: YupMiddlewareOptions;
  validationSchema:
    | Yup.ObjectSchema<_TArgs>
    | ((
        root: _TSource,
        args: _TArgs,
        context: _TContext,
        info: GraphQLResolveInfo,
      ) => Yup.ObjectSchema<_TArgs>);
}
