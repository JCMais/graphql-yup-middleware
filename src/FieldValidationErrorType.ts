import {
  GraphQLObjectType,
  GraphQLString,
  GraphQLNonNull,
  GraphQLList,
} from 'graphql';

export const FieldValidationErrorType = new GraphQLObjectType({
  name: 'FieldValidationError',
  fields: {
    field: { type: new GraphQLNonNull(GraphQLString) },
    errors: {
      type: new GraphQLNonNull(
        new GraphQLList(new GraphQLNonNull(GraphQLString)),
      ),
    },
  },
});

export const FieldValidationError = `
  type FieldValidationError {
    field: String!
    errors: [String!]!
  }
`;
