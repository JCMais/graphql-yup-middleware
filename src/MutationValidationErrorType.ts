import {
  GraphQLString,
  GraphQLList,
  GraphQLObjectType,
  GraphQLNonNull,
} from 'graphql';

import { FieldValidationErrorType } from './FieldValidationErrorType';

export const MutationValidationErrorType = new GraphQLObjectType({
  name: 'MutationValidationError',
  fields: {
    message: { type: new GraphQLNonNull(GraphQLString) },
    details: {
      type: new GraphQLNonNull(
        new GraphQLList(new GraphQLNonNull(FieldValidationErrorType)),
      ),
    },
  },
});

export const MutationValidationError = `
  type MutationValidationError {
    message: String!
    details: [FieldValidationError!]!
  }
`;
