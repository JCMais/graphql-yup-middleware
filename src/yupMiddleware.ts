import { ValidationError } from 'yup';
import {
  GraphQLString,
  GraphQLObjectType,
  isObjectType,
  GraphQLNonNull,
  GraphQLResolveInfo,
} from 'graphql';
import { IMiddleware } from 'graphql-middleware';

import './graphql.augmented';
import buildErrorObjectFromValidationError from './buildErrorObjectFromValidationError';
import { YupMiddlewareOptions, GraphQLExtensionsYup } from './types';
import { MutationValidationErrorType } from './MutationValidationErrorType';

const defaultOptions = {
  errorPayloadBuilder: buildErrorObjectFromValidationError,
  shouldTransformArgs: true,
  yupOptions: {
    abortEarly: false,
  },
};

export default function yupMutationMiddleware<
  TSource = any,
  TContext = any,
  TArgs = any
>(options: YupMiddlewareOptions = {}): IMiddleware<TSource, TContext, TArgs> {
  const hasSuppliedErrorPayloadBuilder = !!options.errorPayloadBuilder;
  const mergedOptions = {
    ...defaultOptions,
    ...options,
  };

  if (typeof options === 'function') {
    throw new TypeError(
      'You have to call the yupMutationMiddleware before adding it to the middlewares: yupMutationMiddleware()',
    );
  }

  return {
    // tslint:disable-next-line: function-name
    async Mutation(
      resolve: Function,
      root: TSource,
      args: TArgs,
      context: TContext,
      info: GraphQLResolveInfo,
    ) {
      const mutationField = info.schema.getMutationType();

      // this should not really happen
      if (!mutationField) {
        return;
      }

      const mutationDefinition = mutationField.getFields()[info.fieldName];

      const hasExtensionsField =
        !!mutationDefinition.extensions &&
        !!mutationDefinition.extensions.yupMiddleware;

      const mutationValidationSchema = hasExtensionsField
        ? (mutationDefinition.extensions!.yupMiddleware as GraphQLExtensionsYup<
            TSource,
            TContext,
            TArgs
          >).validationSchema
        : mutationDefinition.validationSchema;
      const mutationValidationOptions = hasExtensionsField
        ? (mutationDefinition.extensions!.yupMiddleware as GraphQLExtensionsYup<
            TSource,
            TContext,
            TArgs
          >).validationOptions
        : mutationDefinition.validationOptions;

      if (mutationValidationSchema) {
        const finalOptions = {
          ...mergedOptions,
          ...mutationValidationOptions,
        };

        const schema =
          typeof mutationValidationSchema === 'function'
            ? mutationValidationSchema(root, args, context, info)
            : mutationValidationSchema;

        try {
          const values = await schema.validate(args, {
            abortEarly: false,
            ...finalOptions.yupOptions,
          });

          return resolve(
            root,
            finalOptions.shouldTransformArgs ? values : args,
            context,
            info,
          );
        } catch (error) {
          if (error instanceof ValidationError) {
            const errorResult = finalOptions.errorPayloadBuilder(error, {
              root,
              args,
              context,
              info,
            });

            if (!hasSuppliedErrorPayloadBuilder) {
              let returnType = info.returnType;

              // non nullable
              if (info.returnType instanceof GraphQLNonNull) {
                returnType = info.returnType.ofType;
              }

              const isObjReturnType = isObjectType(returnType);

              if (!isObjReturnType) {
                throw new Error(
                  'Only mutations with object return type are supported',
                );
              }

              // returnType cannot be anything else at this point
              const fields = (returnType as GraphQLObjectType).getFields();

              if (!fields.error) {
                throw new Error(
                  'You must have an error field on the payload of your mutation when using default error builder',
                );
              }

              if (fields.error && fields.error.type === GraphQLString) {
                return {
                  error: error.message,
                };
              }

              if (!isObjectType(fields.error.type)) {
                throw new Error(
                  'Mutation payload error field must be of type GraphQLString or GraphQLObjectType',
                );
              }

              // @TODO Improve this validation
              if (fields.error.type.name !== MutationValidationErrorType.name) {
                throw new Error(
                  'Mutation payload error must be of type MutationValidationErrorType if using default options',
                );
              }
            }

            return errorResult;
          } else {
            throw error;
          }
        }
      }

      return resolve(root, args, context, info);
    },
  };
}
