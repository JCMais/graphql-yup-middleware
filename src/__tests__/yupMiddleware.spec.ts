import { graphql } from 'graphql';
import { applyMiddleware } from 'graphql-middleware';
import { makeExecutableSchema } from 'graphql-tools';
import * as yup from 'yup';

import yupMiddleware from '../yupMiddleware';
import { FieldValidationError } from '../FieldValidationErrorType';
import { MutationValidationError } from '../MutationValidationErrorType';
import { YupMiddlewareErrorContext } from '../types';

// got it throwing a dice ðŸŽ²
const randomId = 3;

const getDefaultSchema = <TSource = any, TArgs = Object>(
  validationSchema?: yup.Schema<TArgs>,
) => {
  const typeDefs = `
    type User {
      id: Int!
      firstName: String!
      lastName: String!
      age: Int!
    }
    type AddUserPayload {
      error: String
      user: User
    }
    type AddUserPayloadWithErrorObject {
      error: MutationValidationError
      user: User
    }
    type AddUserPayloadWithErrorCustom {
      errors: [String!]
      user: User
    }
    type Mutation {
      AddUser(firstName: String!, lastName: String!, age: Int!): AddUserPayload!
      AddUserErrorObject(firstName: String!, lastName: String!, age: Int!): AddUserPayloadWithErrorObject!
      AddUserErrorCustom(firstName: String!, lastName: String!, age: Int!): AddUserPayloadWithErrorCustom!
      AddUserWithOptions(firstName: String!, lastName: String!, age: Int!): AddUserPayload!
    }
    type Query {
      hello: String!
    }
  `;

  const AddUser = {
    validationSchema,
    resolve: (_: TSource, args: TArgs) => {
      return {
        user: {
          ...(args as Object),
          id: randomId,
        },
      };
    },
  };

  const resolvers = {
    Mutation: {
      AddUser,
      AddUserErrorObject: AddUser,
      AddUserErrorCustom: AddUser,
      AddUserWithOptions: {
        ...AddUser,
        validationOptions: {
          yupOptions: {
            abortEarly: true,
          },
        },
      },
    },
    Query: {
      hello: () => 'world',
    },
  };

  return makeExecutableSchema({
    typeDefs: [typeDefs, MutationValidationError, FieldValidationError],
    resolvers,
  });
};

const defaultQuery = `
  mutation AddUser($firstName: String!, $lastName: String!, $age: Int!) {
    AddUser(firstName: $firstName, lastName: $lastName, age: $age) {
      error
      user {
        id
        firstName
        lastName
        age
      }
    }
  }
`;

const defaultQueryErrorObject = `
  mutation AddUser($firstName: String!, $lastName: String!, $age: Int!) {
    AddUserErrorObject(firstName: $firstName, lastName: $lastName, age: $age) {
      error {
        message
        details {
          field
          errors
        }
      }
      user {
        id
        firstName
        lastName
        age
      }
    }
  }
`;

const defaultQueryErrorCustom = `
  mutation AddUser($firstName: String!, $lastName: String!, $age: Int!) {
    AddUserErrorCustom(firstName: $firstName, lastName: $lastName, age: $age) {
      errors
      user {
        id
        firstName
        lastName
        age
      }
    }
  }
`;

const defaultQueryWithOptions = `
  mutation AddUser($firstName: String!, $lastName: String!, $age: Int!) {
    AddUserWithOptions(firstName: $firstName, lastName: $lastName, age: $age) {
      error
      user {
        id
        firstName
        lastName
        age
      }
    }
  }
`;

const customErrorPayloadBuilder = (
  error: yup.ValidationError,
  errorContext: YupMiddlewareErrorContext,
) => {
  const reduceErrors = (error: yup.ValidationError): string[] =>
    error.inner.reduce(
      (acc, innerError) => [...acc, ...reduceErrors(innerError)],
      [error.message],
    );
  return {
    errors: reduceErrors(error),
  };
};

it('should validate correctly - string error - error', async () => {
  const schema = getDefaultSchema(
    yup.object().shape({
      firstName: yup
        .string()
        .trim()
        .min(1),
      lastName: yup
        .string()
        .trim()
        .min(1),
      age: yup
        .number()
        .min(18)
        .max(100),
    }),
  );

  const schemaWithMiddleware = applyMiddleware(schema, yupMiddleware());

  const res = await graphql(schemaWithMiddleware, defaultQuery, null, null, {
    firstName: '',
    lastName: '',
    age: 10,
  });

  expect(res).toMatchSnapshot();
});

it('should validate correctly - string error - pass', async () => {
  const schema = getDefaultSchema(
    yup.object().shape({
      firstName: yup
        .string()
        .trim()
        .min(1),
      lastName: yup
        .string()
        .trim()
        .min(1),
      age: yup
        .number()
        .min(18)
        .max(100),
    }),
  );

  const schemaWithMiddleware = applyMiddleware(schema, yupMiddleware());

  const res = await graphql(schemaWithMiddleware, defaultQuery, null, null, {
    firstName: 'Jon',
    lastName: 'Doe',
    age: 18,
  });

  expect(res).toMatchSnapshot();
});

it('should validate correctly - object error - error', async () => {
  const schema = getDefaultSchema(
    yup.object().shape({
      firstName: yup
        .string()
        .trim()
        .min(1),
      lastName: yup
        .string()
        .trim()
        .min(1),
      age: yup
        .number()
        .min(18)
        .max(100),
    }),
  );

  const schemaWithMiddleware = applyMiddleware(schema, yupMiddleware());

  const res = await graphql(
    schemaWithMiddleware,
    defaultQueryErrorObject,
    null,
    null,
    {
      firstName: '',
      lastName: '',
      age: 10,
    },
  );

  expect(res).toMatchSnapshot();
});

it('should validate correctly - object error - pass', async () => {
  const schema = getDefaultSchema(
    yup.object().shape({
      firstName: yup
        .string()
        .trim()
        .min(1),
      lastName: yup
        .string()
        .trim()
        .min(1),
      age: yup
        .number()
        .min(18)
        .max(100),
    }),
  );

  const schemaWithMiddleware = applyMiddleware(schema, yupMiddleware());

  const res = await graphql(
    schemaWithMiddleware,
    defaultQueryErrorObject,
    null,
    null,
    {
      firstName: 'Jon',
      lastName: 'Doe',
      age: 18,
    },
  );

  expect(res).toMatchSnapshot();
});

it('should validate correctly - custom error - error', async () => {
  const schema = getDefaultSchema(
    yup.object().shape({
      firstName: yup
        .string()
        .trim()
        .min(1),
      lastName: yup
        .string()
        .trim()
        .min(1),
      age: yup
        .number()
        .min(18)
        .max(100),
    }),
  );

  const schemaWithMiddleware = applyMiddleware(
    schema,
    yupMiddleware({
      errorPayloadBuilder: customErrorPayloadBuilder,
    }),
  );

  const res = await graphql(
    schemaWithMiddleware,
    defaultQueryErrorCustom,
    null,
    null,
    {
      firstName: '',
      lastName: '',
      age: 10,
    },
  );

  expect(res).toMatchSnapshot();
});

it('should validate correctly - custom error - pass', async () => {
  const schema = getDefaultSchema(
    yup.object().shape({
      firstName: yup
        .string()
        .trim()
        .min(1),
      lastName: yup
        .string()
        .trim()
        .min(1),
      age: yup
        .number()
        .min(18)
        .max(100),
    }),
  );

  const schemaWithMiddleware = applyMiddleware(
    schema,
    yupMiddleware({
      errorPayloadBuilder: customErrorPayloadBuilder,
    }),
  );

  const res = await graphql(
    schemaWithMiddleware,
    defaultQueryErrorCustom,
    null,
    null,
    {
      firstName: 'Jon',
      lastName: 'Doe',
      age: 18,
    },
  );

  expect(res).toMatchSnapshot();
});

it('should do nothing if there are no validation schema', async () => {
  const schema = getDefaultSchema();

  const schemaWithMiddleware = applyMiddleware(schema, yupMiddleware());

  const res = await graphql(schemaWithMiddleware, defaultQuery, null, null, {
    firstName: 'J',
    lastName: 'D',
    age: 10,
  });

  expect(res.data).toBeDefined();
  expect(res.data!.AddUser.error).toBeNull();
  expect(res.data!.AddUser.user.id).toBe(randomId);
});

it('should give error if not using middleware correctly', async () => {
  const schema = getDefaultSchema();

  // @ts-ignore
  const schemaWithMiddleware = applyMiddleware(schema, yupMiddleware);

  const res = await graphql(schemaWithMiddleware, defaultQuery, null, null, {
    firstName: '',
    lastName: '',
    age: 10,
  });

  expect(res.errors).toMatchSnapshot();
});

describe('Options', () => {
  it('should transform args only when specified', async () => {
    const schema = () =>
      getDefaultSchema(
        yup.object().shape({
          firstName: yup
            .string()
            .trim()
            .min(1),
          lastName: yup
            .string()
            .trim()
            .min(1),
          age: yup
            .number()
            .min(18)
            .max(100),
        }),
      );

    const schemaWithMiddleware1 = applyMiddleware(
      schema(),
      yupMiddleware({
        shouldTransformArgs: false,
      }),
    );

    const schemaWithMiddleware2 = applyMiddleware(
      schema(),
      yupMiddleware({
        shouldTransformArgs: true,
      }),
    );

    const variables = {
      firstName: ' Jon ',
      lastName: ' Doe ',
      age: 18,
    };

    const res1 = await graphql(
      schemaWithMiddleware1,
      defaultQuery,
      null,
      null,
      variables,
    );
    expect(res1).toMatchSnapshot();

    const res2 = await graphql(
      schemaWithMiddleware2,
      defaultQuery,
      null,
      null,
      variables,
    );
    expect(res2).toMatchSnapshot();
  });

  it('should forward yup options correctly', async () => {
    const schema = getDefaultSchema(
      yup.object().shape({
        firstName: yup
          .string()
          .trim()
          .min(1),
        lastName: yup
          .string()
          .trim()
          .min(1),
        age: yup
          .number()
          .min(18)
          .max(100),
      }),
    );

    const schemaWithMiddleware = applyMiddleware(
      schema,
      yupMiddleware({
        yupOptions: {
          abortEarly: true,
        },
      }),
    );

    const res = await graphql(schemaWithMiddleware, defaultQuery, null, null, {
      firstName: '',
      lastName: '',
      age: 10,
    });

    expect(res).toMatchSnapshot();
  });

  it('should use validationOptions from mutation definition if present', async () => {
    const schema = getDefaultSchema(
      yup.object().shape({
        firstName: yup
          .string()
          .trim()
          .min(1),
        lastName: yup
          .string()
          .trim()
          .min(1),
        age: yup
          .number()
          .min(18)
          .max(100),
      }),
    );

    const schemaWithMiddleware = applyMiddleware(schema, yupMiddleware());

    const res = await graphql(
      schemaWithMiddleware,
      defaultQueryWithOptions,
      null,
      null,
      {
        firstName: '',
        lastName: '',
        age: 10,
      },
    );

    expect(res).toMatchSnapshot();
  });
});
