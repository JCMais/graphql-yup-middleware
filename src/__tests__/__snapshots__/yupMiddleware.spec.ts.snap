// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options should forward yup options correctly 1`] = `
Object {
  "data": Object {
    "AddUser": Object {
      "error": "age must be greater than or equal to 18",
      "user": null,
    },
  },
}
`;

exports[`Options should transform args only when specified 1`] = `
Object {
  "data": Object {
    "AddUser": Object {
      "error": null,
      "user": Object {
        "age": 18,
        "firstName": " Jon ",
        "id": 3,
        "lastName": " Doe ",
      },
    },
  },
}
`;

exports[`Options should transform args only when specified 2`] = `
Object {
  "data": Object {
    "AddUser": Object {
      "error": null,
      "user": Object {
        "age": 18,
        "firstName": "Jon",
        "id": 3,
        "lastName": "Doe",
      },
    },
  },
}
`;

exports[`Options should use validationOptions from mutation definition if present 1`] = `
Object {
  "data": Object {
    "AddUserWithOptions": Object {
      "error": "age must be greater than or equal to 18",
      "user": null,
    },
  },
}
`;

exports[`should give error if not using middleware correctly 1`] = `
Array [
  [GraphQLError: You have to call the yupMutationMiddleware before adding it to the middlewares: yupMutationMiddleware()],
]
`;

exports[`should validate correctly - custom error - error 1`] = `
Object {
  "data": Object {
    "AddUserErrorCustom": Object {
      "errors": Array [
        "3 errors occurred",
        "firstName must be at least 1 characters",
        "lastName must be at least 1 characters",
        "age must be greater than or equal to 18",
      ],
      "user": null,
    },
  },
}
`;

exports[`should validate correctly - custom error - pass 1`] = `
Object {
  "data": Object {
    "AddUserErrorCustom": Object {
      "errors": null,
      "user": Object {
        "age": 18,
        "firstName": "Jon",
        "id": 3,
        "lastName": "Doe",
      },
    },
  },
}
`;

exports[`should validate correctly - extensions field - object error - error 1`] = `
Object {
  "data": Object {
    "AddUserErrorObjectExtensions": Object {
      "error": Object {
        "details": Array [
          Object {
            "errors": Array [
              "firstName must be at least 1 characters",
            ],
            "field": "firstName",
          },
          Object {
            "errors": Array [
              "lastName must be at least 1 characters",
            ],
            "field": "lastName",
          },
          Object {
            "errors": Array [
              "age must be greater than or equal to 18",
            ],
            "field": "age",
          },
        ],
        "message": "3 errors occurred",
      },
      "user": null,
    },
  },
}
`;

exports[`should validate correctly - extensions field - object error - pass 1`] = `
Object {
  "data": Object {
    "AddUserErrorObjectExtensions": Object {
      "error": null,
      "user": Object {
        "age": 18,
        "firstName": "Jon",
        "id": 3,
        "lastName": "Doe",
      },
    },
  },
}
`;

exports[`should validate correctly - object error - error 1`] = `
Object {
  "data": Object {
    "AddUserErrorObject": Object {
      "error": Object {
        "details": Array [
          Object {
            "errors": Array [
              "firstName must be at least 1 characters",
            ],
            "field": "firstName",
          },
          Object {
            "errors": Array [
              "lastName must be at least 1 characters",
            ],
            "field": "lastName",
          },
          Object {
            "errors": Array [
              "age must be greater than or equal to 18",
            ],
            "field": "age",
          },
        ],
        "message": "3 errors occurred",
      },
      "user": null,
    },
  },
}
`;

exports[`should validate correctly - object error - pass 1`] = `
Object {
  "data": Object {
    "AddUserErrorObject": Object {
      "error": null,
      "user": Object {
        "age": 18,
        "firstName": "Jon",
        "id": 3,
        "lastName": "Doe",
      },
    },
  },
}
`;

exports[`should validate correctly - string error - error 1`] = `
Object {
  "data": Object {
    "AddUser": Object {
      "error": "3 errors occurred",
      "user": null,
    },
  },
}
`;

exports[`should validate correctly - string error - pass 1`] = `
Object {
  "data": Object {
    "AddUser": Object {
      "error": null,
      "user": Object {
        "age": 18,
        "firstName": "Jon",
        "id": 3,
        "lastName": "Doe",
      },
    },
  },
}
`;
